connector/src/main/java/org.apache.flink/connector/nebula/sink/NebulaBatchOutputFormat.java

```diff
@@ -116,49 +106,107 @@ public void open(int i, int i1) throws IOException {
         }
     }
 
+    private void checkCommitException() {
+        if (commitException != null) {
+            throw new RuntimeException("commit records failed", commitException);
+        }
+    }
+
+    private void renewSession() throws IOException {
+        if (session != null) {
+            session.release();
+            session = null;
+        }
+        try {
+            session = nebulaPool.getSession(graphProvider.getUserName(),
+                    graphProvider.getPassword(), true);
+        } catch (NotValidConnectionException | AuthFailedException
+                 | ClientServerIncompatibleException | IOErrorException e) {
+            LOG.error("failed to get graph session", e);
+            throw new IOException("get graph session error", e);
+        }
+        ResultSet resultSet;
+        try {
+            resultSet = session.execute("USE " + executionOptions.getGraphSpace());
+        } catch (IOErrorException e) {
+            LOG.error("switch space error", e);
+            throw new IOException("switch space error", e);
+        }
+        if (!resultSet.isSucceeded()) {
+            LOG.error("switch space failed: " + resultSet.getErrorMessage());
+            throw new IOException("switch space failed: " + resultSet.getErrorMessage());
+        }
+    }
+
     protected abstract NebulaBatchExecutor<T> createNebulaBatchExecutor();
 
     /**
      * write one record to buffer
      */
     @Override
-    public final synchronized void writeRecord(T row) {
+    public final synchronized void writeRecord(T row) throws IOException {
+        checkCommitException();
         nebulaBatchExecutor.addToBatch(row);
+        numPendingRow++;
 
-        if (numPendingRow.incrementAndGet() >= executionOptions.getBatchSize()) {
+        if (executionOptions.getBatchSize() > 0
+                && numPendingRow >= executionOptions.getBatchSize()) {
             commit();
         }
     }
 
     /**
      * commit batch insert statements
      */
-    private synchronized void commit() {
-        String errorExec = nebulaBatchExecutor.executeBatch(session);
-        if (errorExec != null) {
-            errorBuffer.add(errorExec);
+    private synchronized void commit() throws IOException {
+        int maxRetries = executionOptions.getMaxRetries();
+        int retryDelayMs = executionOptions.getRetryDelayMs();
+        boolean failOnError = executionOptions.getFailureHandler().equals(FailureHandlerEnum.FAIL);
+
+        // execute the batch at most `maxRetries + 1` times
+        for (int i = 0; i <= maxRetries; i++) {
+            try {
+                nebulaBatchExecutor.executeBatch(session);
+                numPendingRow = 0;
+                break;
+            } catch (Exception e) {
+                LOG.error(String.format("write data error (attempt %s)", i), e);
```

Comment:
if it's better  to LOG.warn rather than LOG.error?  print the error only when the maxRetries is used out and failOnError is false.
btw, using %d for int i.

Reply:
Good point! I've fixed this.